Do not silently ignore errors during sync

guewen opened the pull request:
If anything wrong happens during a synchronization, the error is silent and the synchronization loop stops. With this fix, the error is logged and the failed sync is retried later.

I propose this pull request on your develop branch as it seems the most advanced one and I'll use it for my current development, but I don't mind to change the target if you want to.

Commit:
Add onErrorRetry parameter

The default action is to silent the error and to break the sync cycle,
so backward compatibility is kept.

If 'onErrorRetry' is set to true, the sync cycle continues whatever
happens during the http call or the during the execution of the
watchers.

'onErrorRetry' could also be a function, allowing to call 'nextSync()'
on the synchronization object to delay a new run of synchronization, by
default after the interval defined in the parameters. 'nextSync()'
possibly takes an interval, which allows to override the default (for
instance we may want to extend the interval when some errors happen).

Commit:
Build dist files

hparfr commented:
Hi,

Currently, the errors are handled on the low level within _handleOdooerrors_
If anything happens, all the callbacks registered in errorInterceptors will be triggered.

Plus, there is no error looping prevention with your code.

hparfr commented:
Have a look here : https://github.com/guewen/angular-odoo/blob/sync-retry-errors/src/components/odoo/jsonRpc-service.js#L274

And here : https://github.com/akretion/prodoo/blob/master/www/js/app.js#L25 for an example.
It will fail show the error and redirect to a stable state: login.

guewen commented:
Okay, I understand that and indeed I clearly see how to use the interceptors to, for instance, redirect to the login if the session is expired.

Now I don't see how this applies for the use case of `syncImportObject`.

Here is my expectation:
I start a `syncImportObject`, add a watcher on it. I expect it to run for instance every 5000ms, so every 5000ms it will make an HTTP request to Odoo. If one of the call fails, or if a watcher throws an error, I expect the `syncImportObject` to retry in 5000ms.

And the current behavior is the following: if one of the occurrence of `syncImportObject` fails, the loop is stopped and no further sync is proceeded.

I don't see how an interceptor could fix that, but I'm merely learning angular so excuse me if what I say is dumb.

guewen commented:
> Plus, there is no error looping prevention with your code.

Could you expand on that? Isn't `syncImportObject` basically a loop? The `stop` callback is still there, I didn't changed the logic apart of the setting a retry on the errors.

hparfr commented:
> I don't see how an interceptor could fix that, but I'm merely learning angular so excuse me if what I say is dumb.

No problem on that !  :)
May be there is better design for this ^^

The interceptor is may be not well named, it's not an implementation of $http.interceptor but it's similar.
If an error happens it will trigger.
Then it's the responsibility of the application (and not the library) to take curative actions.

> And the current behavior is the following: if one of the occurrence of syncImportObject fails, the loop is stopped and no further sync is proceeded.

Exactly

If the server is down, there is no need to query it again and again and it may not be a good thing to continue to accept user input if the link is known to be broken : it depends form app to app and it's the responsibility of the app to handle it.

In the example, the user is redirected to login page because it's the initial state of the app (and the most stable one).

hparfr commented:
> Could you expand on that? Isn't syncImportObject basically a loop? The stop callback is still there, I didn't changed the logic apart of the setting a retry on the errors.

The stop() is available to the consumers (application). 

In the current implementation, the call to stop() is not needed in case of an error because the promise will be rejected and $timeout() will not be called.

guewen commented:
Ok so if I understand, your expectation is that once you have an error, you should stop the loop, is that right? I find it a bit surprising, I was really thinking that it was acting like a cron, if an occurrence of a cron fails, it continues.

> If the server is down, there is no need to query it again and again and it may not be a good thing to continue to accept user input if the link is known to be broken : 

In my use case, it is exactly what I want, because if I start a recurring sync to retrieve the data from odoo, if at a point in time it couldn't get new data, I don't matter if I know that in a next iteration it will get it. We can have many more errors than a crude "server down" error (especially if we are on mobile phones where the network instability is something we have to deal with) and resetting the state of the application to the login on any error is rude.

> it depends form app to app and it's the responsibility of the app to handle it.

I don't see how to achieve that with the current interceptor: the only context it has is the error Object: when the interceptor catches an error, I have no way to know that the error comes from the sync loop and that I should trigger a new sync loop (or if I have a way to do that I miss it).  

> In the current implementation, the call to stop() is not needed in case of an error because the promise will be rejected and $timeout() will not be called.

That's the point, we have different use cases. Would you agree if I find a way to pass a `retry` callback which is a function, the default one doing nothing, so we can pass a function that retries?

guewen commented:
(or simply a boolean flag defining if we retry on errors or not?)

hparfr commented:
Having a retry()  in the sync api or a "continue" flag are good.
It's even possible to implement it without breaking changes.

guewen commented:
I updated my PR with a proposal. Let me know if that's a correct  way to do it:

An implementation examples gives:

``` javascript

.provider('OdooSyncUsers', [function() {

  var sync = {};
  sync.importer = null;

  this.$get = [
    'jsonRpc', '$q', '$log',
    function(jsonRpc, $q, $log) {

    sync.syncData = function(options, objectRef) {

      var sync_setup = {
        model: 'res.users',
        func_key: 'auto',
        domain: [],
        limit: 50,
        interval: 5000,  // 5 seconds
        onError: function(err) {
          $log.error('Synchronization error', err);
          sync.importer.nextSync();
        }
      };

      sync.importer = jsonRpc.syncImportObject(sync_setup);

      sync.importer.watch(function(e) {
        $log.info(objectRef);

      });

      return sync.importer;
    }

    return sync;
  }];

  return this;
}])


```

If that's ok for you, I will squash my commits (BTW should I do the builds or not?).

hparfr commented:
If you want to always nextSync onError, there is no need to expose it through the API :  

``` js
var p = odooRpc.syncDataImport(
            params.model,
            params.func_key,
            params.base_domain,
            params.filter_domain,
            params.limit,
            object)
        .then(nextSync)
        .then(runWatchers);
        if (params.continueOnError)
            p.catch(nextSync);
```

guewen commented on dist/odoo.js:
```diff
@@ -155,14 +178,8 @@ angular.module('odoo').provider('jsonRpc', function jsonRpcProvider() {
-					});
-				});
+				.then(nextSync)
+				.then(runWatchers, params.onError);
```
I goofed my last changes, I'll fix that soon

guewen commented:
The possibility to pass a function gives a more fine-grained control over the errors than a simple Boolean. Examples: log something, call nextSync after a longer $timeout for some types of errors or for repeated errors (counter), stop completely the loop on specific conditions, ...

guewen commented:
I reworked my proposal. I hope you will like it.
I attempted to provide a simple API for simple cases, so you can just toggle retry on errors by setting `onErrorRetry` to `true`, but I also kept the flexibility supplied by the function.

Verbatim of my commit:

> The default action is to silent the error and to break the sync cycle,
> so backward compatibility is kept.
> 
> If 'onErrorRetry' is set to true, the sync cycle continues whatever
> happens during the http call or the during the execution of the
> watchers.
> 
> 'onErrorRetry' could also be a function, allowing to call 'nextSync()'
> on the synchronization object to delay a new run of synchronization, by
> default after the interval defined in the parameters. 'nextSync()'
> possibly takes an interval, which allows to override the default (for
> instance we may want to extend the interval when some errors happen).

Examples of `syncImportObject` params:

Same behavior than now (cycle breaks on errors):

``` js
      var sync_setup = {
        model: 'res.users',
        func_key: 'auto',
        base_domain: [],
        filter_domain: [],
        limit: 50,
        interval: 5000
      };
```

Always retry on errors:

``` js
      var sync_setup = {
        model: 'res.users',
        func_key: 'auto',
        base_domain: [],
        filter_domain: [],
        limit: 50,
        interval: 5000,
        onRetryError: true
      };
```

Retry with a function:

``` js
      var sync_setup = {
        model: 'res.users',
        func_key: 'auto',
        base_domain: [],
        filter_domain: [],
        limit: 50,
        interval: 5000,
        onRetryError: function(sync, err) {
          $log.error('Synchronization error', err);

          // we can for instance have a counter of errors and 
          // do an action/extend the delay after too many retries

          if(shouldExtendDelay(err)) {
            sync.nextSync(1000 * 60 * 15);
          } else {
            // default delay
            sync.nextSync();
          }
        }
      };
```

Commit:
Add a closure function for onErrorRetry

So we have an access to the current sync 'object' directly from the
onErrorRetry args.

hparfr commented:
Thanks Guewen, it's a really great work :)
